### ISSUE TITLE
AI Purchase is not looking at the unplaced units from current & previous turns

### ISSUE THREAD

**Issue by TheDog-GH** [triplea-game/triplea#12436]
I am working on a new map, it has random starting territories, Risk like.  So this is difficult to reproduce, but I am willing to test the following.

Using **FastAI** or **HardAI** the AI is buying Factories even though there are Factories available for placing, just given to it that turn and in previous turns.

The AI Purchase is **not looking** at the unplaced units from current & previous turns that have not been placed.

This property is important
<property name="Unplaced units live when not placed" 	value="true"

Extract from Game History
Trigger NationalTheme Germany-AI: **2 Industry-Hvys, 2 Industry-Lgts and 2 Industry-Meds** gained by PlayerId named:Germany
Germany buy 2 Armor-Infs, 1 Base-Camp, 1 Biplane, 6 Cavalrys, 2 HQ-Armys, **1 Industry-Hvy**, 1 Inf-Elite, 6 Inf-Traineds and 1 _Dev-Armor-Lgt; Remaining resources: 6 PUs; 2 Tech;

**Comment by beelee1** [comment]
Does it do the same in 2.5 ? Might have always been like that. Did you ask wc ? Maybe it's a map making/xml thing ?

Idk :)

**Comment by TheDog-GH** [comment]
Yes it does the same in 2.5 :)

In Risk style maps the AI conquers territory and will **not** place Factories unless it thinks it needs them, here I am trying to emulate what a human would do, buy pre-buying the Factories so as the AI expands it will need Factories, but because it has the Factories it **will** place them.  The AI does not know the Factories generate PU & Tech, but a human does.

For AI purchasing it might be 'messy' to look what is already unplaced and then the AI decides what to purchase.

A big problem with Risk style maps is that, the AI does not always place Factories when a human would.

Maybe another answer is the AI placement has a new phase, at the end of its current AI placement phase & places the best Factory in a Territory.  This way it would use up its stock of Factories and not interfere with the  current AI purchase & place phases?

**Comment by beelee1** [comment]
Right on. Should this be a Feature Request then ?

**Comment by TheDog-GH** [comment]
Maybe, yeah I suppose it should, could you do the honors?

**Comment by DanVanAtta** [comment]
Fixed rolled back, re-opening

**Comment by DanVanAtta** [comment]
Lining issue: https://github.com/triplea-game/triplea/issues/14439, has interesting investigative details.
CC: @VictoryFirst01

**Comment by VictoryFirst2** [comment]
@DanVanAtta Good that you take a look at this again, as I think the original solution was a bit too simple. Although I must say I have spent a lot of time trying to fix the issue that the AI does not place its unplaced factories (discovered because of https://github.com/triplea-game/triplea/issues/14439). I have finally found a solution and was about to upload a pull request. This should have been a fix for https://github.com/triplea-game/triplea/issues/14439 but you closing it and reopening this makes me think you are planning to solve that issue while solving this one. In any case, I will upload my work so that you can review it, and perhaps get some inspiration to create a more elegant solution.

**Comment by frigoref** [comment]
@DanVanAtta You might have missed that @VictoryFirst01 already analyzed issue that unplaced factories are not getting placed (issue #14439) is caused  in the placement step.
This issue here #12436 and the attempted fix PR #14411 therefore didn't need to be reverted by reverting PR #14454  as the fix PR is related to the purchasing step before the placement step.
I'd suggest reintroducing the changes of fix PR #14411 again and letting @VictoryFirst01 provide a new PR to fix issue #14439 .

**Comment by VictoryFirst2** [comment]
@DanVanAtta I agree with @frigoref, as https://github.com/triplea-game/triplea/issues/14439 is a different issue than https://github.com/triplea-game/triplea/issues/12436. https://github.com/triplea-game/triplea/issues/14439 would be: AI is not placing unplaced factories. https://github.com/triplea-game/triplea/issues/12436 would be more like a feature request: AI should more thoroughly consider the units in its unplaced units pool while purchasing new units. Not simply not buying any factories if you just have one in the unplaced units pool, but also not ignoring them completely.

**Comment by DanVanAtta** [comment]
@frigoref / @VictoryFirst01 please kindly break up the issues into clean bug reports each and close down the old ones and link to the new ones. Please consider opening a discussion to explore the overall topic of AI factory placement. 

We really need to step up our bug reporting game. I created this discussion thread to explore this more: https://github.com/triplea-game/triplea/discussions/14456

**Comment by DanVanAtta** [comment]
Responding to @VictoryFirst01 


> Good that you take a look at this again,

No, I'm not. I'm purely doing administrative work to try and keep our bug reporting database (Github Issues) accurate.

>  as I think the original solution was a bit too simple. 

Agree. 

> Although I must say I have spent a lot of time trying to fix the issue

Thank you

> I have finally found a solution and was about to upload a pull request. 

Please do so, if a code update helps, then publish it.

> This should have been a fix for https://github.com/triplea-game/triplea/issues/14439 but you closing it and reopening 

If the problem exists in the HEAD revision of 'main', then the issue ought to be open. Otherwise, closed. I'm working against myself by holding a discussion and Q&A in a BUG REPORT.


> this makes me think you are planning to solve that issue while solving this one.

No, it is the assignee field that indicates an item is being actively worked on. If you can't assign an issue to yourself, please request it be done.

>  In any case, I will upload my work so that you can review it,

Great.


> and perhaps get some inspiration to create a more elegant solution.

Unlikely. My priorities is to get TripleA onto a better release model (which starts with releasing latest HEAD of main)  and trying to delegate as much of the community & project management so that TripleA can be as self-running as possible.

**Comment by VictoryFirst2** [comment]
@DanVanAtta Thanks for clarifying. I will write new detailed bug reports for the two issues and ask @TheDog-GH to close his (or potentially, @TheDog-GH could modify his existing bug reports). EDIT I think the issue here is already explained well enough, so I will only write a new one for the other issue: https://github.com/triplea-game/triplea/issues/14439

I have one more question, what is the "HEAD of main"? What problems belong to this and where can I see them? If these problems have a priority, that would be good to know, as I will try to fix problems belonging to this from now on.

**Comment by VictoryFirst2** [comment]
@DanVanAtta Here is the updated bug report for https://github.com/triplea-game/triplea/issues/14439:

https://github.com/triplea-game/triplea/issues/14457

I will mention this new issue in my fix PR. The old bug report can remain closed.

I will not create a new bug report for the issue in this thread because I think the issue is already explained well enough.
